src/ripple/protocol/TER.h

```diff
@@ -278,7 +282,17 @@ enum TECcodes : TERUnderlyingType {
     tecKILLED = 150,
     tecHAS_OBLIGATIONS = 151,
     tecTOO_SOON = 152,
-    tecHOOK_ERROR [[maybe_unused]] = 153
+    tecHOOK_ERROR [[maybe_unused]] = 153,
+    tecMAX_SEQUENCE_REACHED = 154,
+    tecNO_SUITABLE_PAGE = 155,
+    tecBUY_SELL_MISMATCH = 156,
+    tecOFFER_TYPE_MISMATCH = 157,
+    tecCANT_ACCEPT_OWN_OFFER = 158,
+    tecINSUFFICIENT_FUNDS = 159,
+    tecOBJECT_NOT_FOUND = 160,
```

Comment:
This could be `tecNO_NFT_OFFER` rather than the more generic `OBJECT_NOT_FOUND`. But I don't object, I'm fine either way. I'm making a note because we already have errors like `tecNO_LINE` and `tecNO_DST` that are specific to the missing objects.

Reply:
@nbougalis I have a more general concern about the additional `tec` codes.  The only codes that need to be captured in the metadata are `tes` and `tec`.  From years ago I remember hearing that there is some trick for how `tec` codes are serialized that does not allow us to currently use, say, a full 256 values for possible `tec` codes.  As I recall, vaguely, that limit was around 159.

Is that recollection correct?  If so, how close does this bring us to whatever the current limit is?  Is it time to put an escape hatch in place so we can support new `tec` codes with yet higher numbers?

Reply:
@nbougalis I have a more general concern about the additional `tec` codes.  The only codes that need to be captured in the metadata are `tes` and `tec`.  From years ago I remember hearing that there is some trick for how `tec` codes are serialized that does not allow us to currently use, say, a full 256 values for possible `tec` codes.  As I recall, vaguely, that limit was around 159.

Is that recollection correct?  If so, how close does this bring us to whatever the current limit is?  Is it time to put an escape hatch in place so we can support new `tec` codes with yet higher numbers?

Reply:
I did a little more digging into this question.  Only `tesSUCCESS` ('0x0`) and `tec` codes (`0x64` through `0xFF`) are stored in metadata.  They are captured in an `sfTransactionResult`, which is a `UINT8`.  So without any complicated encoding we can capture a total of 155 `tec` codes.  With this commit we will have used 64 of those codes, so we'll have 91 left -- more than half.

From there we could start using the range from `0x1` through `0x63` for `tec` codes.  After we've used those we can't create more `tec` codes without a more complicated encoding.

So it looks like the need is not urgent.  But it wouldn't hurt to have a plan for `tec` codes beyond 255.

Reply:
Addressing the naming question of `tecOBJECT_NOT_FOUND` vs `tecNO_NFT_OFFER`, my personal bias is toward `tecNO_NFT_OFFER`.  I find failing transactions really difficult to debug because the error codes are often quite unspecific.  We have way too many cases of `temMALFORMED`.

On the flip side, I also see the value minimizing the number of error codes -- especially `tec` codes -- in the code base.  See the previous discussion.

So I'm inclined to leave this code alone to make it more reusable.  Unless other folks weigh in on the side of `tecNO_NFT_OFFER`, in which case I'm fine with changing the name.

Reply:
I looked more closely at the proposed `TER` codes.  There were a few that were defined but unused.  There were a few more that I decided to make more consistent with the API by adding `NFTOKEN` to their strings.  Here's the commit with the changes I'm proposing: https://github.com/ripple/rippled/pull/4101/commits/043afbdd3b22f4c35c87e3566283960e3e35b26b

I think this is an improvement over the current situation, but it also is a change to the API.  I don't believe any of these `TER` codes are part of the spec.  So it's a relatively minor change, as API changes go.  I believe this should have minimal impact on people developing new XLS20 code.

But I've also isolated the commit, so if someone says that we shouldn't make this change it's relatively easy to undo.

Let me know what you think.